### PR TITLE
feat(replays): Parse and emit user click actions to the ingest-replays snuba consumer

### PR DESCRIPTION
**Pull request by cmanallen**
Adds support for indexing click events present in the RRWeb recording data.

1. Decompresses and parses recording data.
2. Iterates over each recording event.
3. Click events are extracted into a collection.
4. The click events are then published to the replays snuba consumer.

Depends on: https://github.com/getsentry/snuba/pull/3855

Relates to: https://github.com/getsentry/sentry/issues/45524

### CODE REVIEW

**Comment by bruno-garcia** [comment]
DIscussed on a call to have this for now only count clicks and emit metrics. We'll get this live in Sentry to see how the parsing behaves and get an idea of rows we'll be writing.

**Review by JoshFerge — COMMENTED** [review]
could we modify this PR to not emit to replays-snuba so we can quickly validate that this does not backlog our consumer at existing scale in production?

**Review by JoshFerge — COMMENTED** [review]
will give further review tomorrow am

**Comment by cmanallen** [comment]
@JoshFerge https://github.com/getsentry/sentry/pull/46315

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+from sentry.utils import json, kafka_config
+from sentry.utils.pubsub import KafkaPublisher
+
+replay_publisher: Optional[KafkaPublisher] = None
+
+ReplayActionsEventPayloadClick = TypedDict(
+    "ReplayActionsEventPayloadClick",
+    {
+        "node_id": int,
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:16]: is `node_id` something people would ever search on? it's internal to RRWeb right?

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    clicks: List[ReplayActionsEventPayloadClick],
+) -> ReplayActionsEventPayload:
+    return {
+        "type": "replay_actions",
+        "replay_id": replay_id,
+        "clicks": clicks,
+    }
+
+
+def iter_user_actions(segment_data: bytes) -> Iterator[ReplayActionsEventPayloadClick]:
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:101]: what's the advantage of returning an iterator here? or could this function just return a list?

Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:101]: given the previous PR #46315 maybe this function could just take in the decompressed and parsed payload?

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                    "id": attributes.get("id", ""),
+                    "class": attributes.get("class", "").split(" "),
+                    "text": node["textContent"],
+                    "role": attributes.get("role", ""),
+                    "alt": attributes.get("alt", ""),
+                    "testid": attributes.get("testid", ""),
+                    "aria_label": attributes.get("aria-label", ""),
+                    "title": attributes.get("title", ""),
+                    "timestamp": int(payload["timestamp"]),
+                    "event_hash": uuid.uuid4().hex,
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:123]: is there any way we could make the event_hash some kind of deterministic value? maybe replay_id + timestamp + node_id? to handle possible duplicate kafka messages given kafka's "at least once" delivery guarantees.

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "replay_id": replay_id,
+        "clicks": clicks,
+    }
+
+
+def iter_user_actions(segment_data: bytes) -> Iterator[ReplayActionsEventPayloadClick]:
+    """Return a list of ReplayActionsEventPayloadClick types."""
+    events = json.loads(decompress(segment_data))
+
+    for event in events:
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:105]: could we add a sentry span for this?

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def _initialize_publisher() -> KafkaPublisher:
+    global replay_publisher
+
+    if replay_publisher is None:
+        config = settings.KAFKA_TOPICS[settings.KAFKA_INGEST_REPLAY_EVENTS]
+        replay_publisher = KafkaPublisher(
+            kafka_config.get_kafka_producer_cluster_options(config["cluster"]),
+            asynchronous=False,
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:142]: i'm wondering if we need to set async here or not.

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            payload = event["data"].get("payload", {})
+            if payload.get("category") == "ui.click":
+                node = payload.get("data", {}).get("node", {})
+                attributes = node.get("attributes", {})
+                yield {
+                    "node_id": node["id"],
+                    "tag": node["tagName"],
+                    "id": attributes.get("id", ""),
+                    "class": attributes.get("class", "").split(" "),
+                    "text": node["textContent"],
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:116]: do we need to check a max string length for textContent? or any of these other fields? are they all unbounded currently? / what are their current limits?

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    retention_days: int,
+    payload: ReplayActionsEventPayload,
+) -> ReplayActionsEvent:
+    return {
+        "type": "replay_event",
+        "start_time": time.time(),
+        "replay_id": replay_id,
+        "project_id": project_id,
+        "retention_days": retention_days,
+        "payload": list(bytes(json.dumps(payload).encode())),  # type: ignore
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:86]: doesn't `encode` already return `bytes`? why do you need to call `bytes` again?

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+from typing import Iterator, List, Literal, Optional, TypedDict
+
+from django.conf import settings
+
+from sentry.utils import json, kafka_config
+from sentry.utils.pubsub import KafkaPublisher
+
+replay_publisher: Optional[KafkaPublisher] = None
+
+ReplayActionsEventPayloadClick = TypedDict(
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:13]: do you have to use the `TypedDict` constructor here?

**Review by JoshFerge — COMMENTED** [review]
looks great! just some small comments / questions

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 1 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py:16]: Its not something we would want to search for but definitely useful data that could be returned to the client.

Comment (JoshFerge) [reply src/sentry/replays/usecases/ingest/dom_index.py:16]: ah good point. makes sense.

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 4 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py:105]: For each item or the whole loop?

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 6 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py:116]: They're bounded in snuba but we should match here as well.

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 8 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py:13]: `class` is a reserved keyword.

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 4 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py:105]: I think this is mostly accomplished already by the timer in `parse_and_emit_replay_actions`.  It does collect more than just the loop (some dictionary assignments and a publish to kafka - which is async).  But the non-loop operations are very small and unlikely to strongly influence the result.  Thoughts on using just the single outer metrics timer?

**Review by JoshFerge — APPROVED** [review]

File: src/sentry/replays/usecases/ingest/dom_index.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+
+def get_user_actions(
+    replay_id: str,
+    events: List[Dict[str, Any]],
+) -> List[ReplayActionsEventPayloadClick]:
+    """Return a list of ReplayActionsEventPayloadClick types."""
+    result: List[ReplayActionsEventPayloadClick] = []
+    for event in events[:EVENT_LIMIT]:
```
Comment (JoshFerge) [inline src/sentry/replays/usecases/ingest/dom_index.py:110]: don't we want this to be the number of click events, not the total number of events in the payload? I think this wouldn't work very well as there are usually a lot in one segment

**Review by JoshFerge — CHANGES_REQUESTED** [review]

File: src/sentry/replays/usecases/ingest/dom_index.py (hunk 9 above)
Comment (cmanallen) [reply src/sentry/replays/usecases/ingest/dom_index.py]: Oof.  Thanks for catching that :D

**Review by JoshFerge — APPROVED** [review]

**Pull request merged**